### PR TITLE
Reform opid

### PR DESCRIPTION
**Pull request by subrotomukherjee**
**Before creating a pull request make sure that:**
PAY-2874

Add OIDC support to Fee Register

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: dependency-check-suppressions.xml [hunk 1]
```diff
@@ -43,4 +43,8 @@
     <suppress>
         <cve>CVE-2020-5398</cve>
     </suppress>
+
```
Comment (YasharRahvar) [inline dependency-check-suppressions.xml:46]: try not to have any cve

File: api/src/functionalTest/java/uk/gov/hmcts/fees/register/functional/idam/IdamApi.java [hunk 2]
```diff
@@ -64,13 +58,16 @@ TokenExchangeResponse exchangeCode(@Param("code") String code,
...
-        private String code;
+        @JsonProperty("id_token")
+        private String idToken;
+
+        @JsonProperty("access_token")
+        private String accessToken;
     }
 
-    @Data
+    /*@Data
```
[the remark this answers is not in the document]
Comment (subrotomukherjee) [reply api/src/functionalTest/java/uk/gov/hmcts/fees/register/functional/idam/IdamApi.java:68]: done

File: api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/AuthCheckerConfiguration.java [hunk 3]
```diff
@@ -1,21 +1,30 @@
...
 
 @Configuration
+@Lazy
+@EnableFeignClients(basePackageClasses = {IdamApi.class})
 public class AuthCheckerConfiguration {
 
     @Bean
     public Function<HttpServletRequest, Collection<String>> authorizedRolesExtractor() {
-        return (any) -> Collections.emptyList();
+        return (any) -> Stream.of("freg-editor", "freg-approver", "freg-admin").collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (subrotomukherjee) [reply api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/AuthCheckerConfiguration.java:27]: done

File: api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java [hunk 4]
```diff
@@ -3,30 +3,63 @@
...
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Optional;
+import java.util.function.Function;
 
 import static org.springframework.security.config.http.SessionCreationPolicy.STATELESS;
 
 @Configuration
-@EnableWebSecurity
+@EnableWebSecurity(debug = true)
```
[the remark this answers is not in the document]
Comment (subrotomukherjee) [reply api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java:38]: done

File: api/src/main/java/uk/gov/hmcts/fees/register/api/security/model/UserDetails.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package uk.gov.hmcts.fees.register.api.security.model;
+
+import java.util.Collection;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+
+import static java.util.stream.Collectors.toList;
+
+public class UserDetails extends org.springframework.security.core.userdetails.User {
```
[the remark this answers is not in the document]
Comment (subrotomukherjee) [reply api/src/main/java/uk/gov/hmcts/fees/register/api/security/model/UserDetails.java:9]: done

File: api/src/main/resources/application.properties [hunk 6]
```diff
@@ -1,6 +1,8 @@
 spring.application.name=fees-register-api
 spring.jackson.serialization.fail-on-unwrapped-type-identifiers=false
 
+logging.level.org.springframework.security=DEBUG
```
[the remark this answers is not in the document]
Comment (subrotomukherjee) [reply api/src/main/resources/application.properties:4]: done.

File: api/src/main/resources/application.properties [hunk 7]
```diff
@@ -22,3 +24,15 @@ management.endpoint.health.show-details=always
...
+
+idam.api.url=${idam_api_url:https://idam-api.aat.platform.hmcts.net}
+
+spring.security.oauth2.client.provider.oidc.issuer-uri=${ISSUER_URI:https://idam-web-public.aat.platform.hmcts.net/o}
+
+oidc.audience-list = fees_admin_frontend
+
+oidc.issuer = ${OIDC_ISSUER:https://forgerock-am.service.core-compute-idam-aat2.internal:8443/openam/oauth2/hmcts}
+
+#base-urls.idam=${IDAM_API_URL:http://localhost:5000}
```
[the remark this answers is not in the document]
Comment (subrotomukherjee) [reply api/src/main/resources/application.properties:38]: done

File: api/src/test/java/uk/gov/hmcts/fees/register/api/componenttests/backdoors/IdamRepositoryMock.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+    private void populateSecurityContextHolder(ImmutableList<String> roles/*, Authentication auth*/) {
+        final List<RoleMock> authorities = new ArrayList<>();
+
+        for (final String role : roles) {
+            final RoleMock r = new RoleMock();
+            r.setAuthority(role);
+            authorities.add(r);
+        }
+        //SecurityContextHolder.getContext().setAuthentication(auth);
```
[the remark this answers is not in the document]
Comment (subrotomukherjee) [reply api/src/test/java/uk/gov/hmcts/fees/register/api/componenttests/backdoors/IdamRepositoryMock.java:54]: Done

File: api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java [hunk 9]
```diff
@@ -37,28 +70,56 @@ public void configure(WebSecurity web) {
             "/v2/**",
             "/health",
             "/health/liveness",
-            "/info");
+            "/info","/error");
```
Comment (amitchidrewar1301) [inline api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java:73]: Please let us know what this been added

File: api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java [hunk 10]
```diff
@@ -37,28 +70,56 @@ public void configure(WebSecurity web) {
...
             .antMatchers(HttpMethod.DELETE, "/fees-register/fees/**", "/fees/**/versions/**").hasAnyAuthority("freg-editor", "freg-admin")
-            .antMatchers(HttpMethod.GET, "/fees-register/fees/**").permitAll();
+            .antMatchers(HttpMethod.GET, "/fees-register/fees/**").permitAll()
+            .and()
+            .oauth2ResourceServer()
+            .jwt()
+            .jwtAuthenticationConverter(jwtAuthenticationConverter)
+            .and()
+            .and()
+            /*.oauth2Client()*/;
```
Comment (amitchidrewar1301) [inline api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java:100]: Please uncomment line no. 100, its as per reference from open id specs

File: api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java [hunk 11]
```diff
@@ -37,28 +70,56 @@ public void configure(WebSecurity web) {
...
+    }
+
+    @Bean
+    public static PropertySourcesPlaceholderConfigurer propertyPlaceHolderConfigurer() {
```
Comment (amitchidrewar1301) [inline api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java:104]: Please check if you can remove this bean definition

File: api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/security/converter/FRJwtGrantedAuthoritiesConverter.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        }
+        return authorities;
+    }
+
+    /**
+     * Method responsible to get stream of authorities based on claims
+     * @param roles
+     * @return
+     */
+    private List<GrantedAuthority> extractAuthorityFromClaims(List<String> roles) {
```
Comment (amitchidrewar1301) [inline api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/security/converter/FRJwtGrantedAuthoritiesConverter.java:59]: Can we remove if not used

File: api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java (hunk 10 above)
Comment (subrotomukherjee) [reply api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java:100]: done

File: api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/security/converter/FRJwtGrantedAuthoritiesConverter.java (hunk 12 above)
Comment (subrotomukherjee) [reply api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/security/converter/FRJwtGrantedAuthoritiesConverter.java:59]: This is required. The method name was a misnomer. I have changed it.

File: api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java (hunk 9 above)
Comment (subrotomukherjee) [reply api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java]: This has been removed now.

File: api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java (hunk 11 above)
Comment (subrotomukherjee) [reply api/src/main/java/uk/gov/hmcts/fees/register/api/configuration/SpringSecurityConfiguration.java]: Done. The code has been removed now.

File: api/src/test/java/uk/gov/hmcts/fees/register/api/componenttests/sugar/RestActions.java [hunk 13]
```diff
@@ -5,31 +5,31 @@
...
 import java.util.UUID;
 
 import static org.springframework.http.MediaType.APPLICATION_JSON;
-import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
 public class RestActions {
     private final HttpHeaders httpHeaders = new HttpHeaders();
     private final MockMvc mvc;
-    private final UserResolverBackdoor userRequestAuthorizer;
+    //private final UserResolverBackdoor userRequestAuthorizer;
```
[the remark this answers is not in the document]
Comment (subrotomukherjee) [reply api/src/test/java/uk/gov/hmcts/fees/register/api/componenttests/sugar/RestActions.java]: Done.

File: api/src/functionalTest/java/uk/gov/hmcts/fees/register/functional/idam/IdamService.java [hunk 14]
```diff
@@ -9,7 +9,7 @@
 import uk.gov.hmcts.fees.register.functional.config.TestConfigProperties;
 import uk.gov.hmcts.fees.register.functional.idam.IdamApi.CreateUserRequest;
 import uk.gov.hmcts.fees.register.functional.idam.IdamApi.Role;
-import uk.gov.hmcts.fees.register.functional.idam.IdamApi.TokenExchangeResponse;
+//import uk.gov.hmcts.fees.register.functional.idam.IdamApi.TokenExchangeResponse;
```
[the remark this answers is not in the document]
Comment (subrotomukherjee) [reply api/src/functionalTest/java/uk/gov/hmcts/fees/register/functional/idam/IdamService.java]: Done.

File: dependency-check-suppressions.xml (hunk 1 above)
Comment (amitchidrewar1301) [reply dependency-check-suppressions.xml:46]: Please add latest nimbus jar, and remove CVE suppression 17195 , refer commit  https://github.com/hmcts/ccpay-bulkscanning-app/pull/195/files#diff-c197962302397baf3a4cc36463dce5eaR222

**Pull request closed without merging**